node/src/chain/acceptor.rs

```diff
@@ -0,0 +1,292 @@
+    /// Most recently accepted block a.k.a blockchain tip
+    mrb: RwLock<Block>,
+
+    /// List of provisioners of actual round
+    provisioners_list: RwLock<Provisioners>,
+
+    /// Upper layer consensus task
+    task: RwLock<super::consensus::Task>,
+
+    db: Arc<RwLock<DB>>,
+    vm: Arc<RwLock<VM>>,
+    network: Arc<RwLock<N>>,
+}
+
+impl<DB: database::DB, VM: vm::VMExecution, N: Network> Drop
+    for Acceptor<N, DB, VM>
+{
+    fn drop(&mut self) {
+        if let Ok(mut t) = self.task.try_write() {
+            t.abort()
+        }
+    }
+}
+
+impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
+    pub async fn new_with_run(
+        keys_path: &str,
+        mrb: &Block,
+        provisioners_list: &Provisioners,
+        db: Arc<RwLock<DB>>,
+        network: Arc<RwLock<N>>,
+        vm: Arc<RwLock<VM>>,
+    ) -> Self {
+        let mut acc = Self {
+            mrb: RwLock::new(mrb.clone()),
+            provisioners_list: RwLock::new(provisioners_list.clone()),
+            db: db.clone(),
+            vm: vm.clone(),
+            network: network.clone(),
+            task: RwLock::new(Task::new_with_keys(keys_path.to_owned())),
+        };
+
+        acc.task.write().await.spawn(
+            &mrb.header,
+            &provisioners_list.clone(),
+            &db,
+            &vm,
+            &network,
+        );
+
+        acc
+    }
+    // Re-route message to consensus task
+    pub(crate) async fn reroute_msg(&self, msg: Message) {
+        match &msg.payload {
+            Payload::NewBlock(_) | Payload::Reduction(_) => {
+                self.task.read().await.main_inbound.send(msg).await;
+            }
+            Payload::Agreement(_) | Payload::AggrAgreement(_) => {
+                self.task.read().await.agreement_inbound.send(msg).await;
+            }
+            _ => warn!("invalid inbound message"),
+        }
+    }
+
+    pub(crate) async fn try_accept_block(
+        &self,
+        blk: &Block,
+        enable_consensus: bool,
+    ) -> anyhow::Result<()> {
+        let mut task = self.task.write().await;
+        let (_, public_key) = task.keys.clone();
+
+        // Verify Block Header
+        self.verify_block_header(&public_key, &blk.header).await?;
+
+        let mut mrb = self.mrb.write().await;
+        let mut provisioners_list = self.provisioners_list.write().await;
+
+        // Reset Consensus
+        task.abort_with_wait().await;
+
+        // Persist block in consistency with the VM state update
+        {
+            let vm = self.vm.write().await;
+            self.db.read().await.update(|t| {
+                t.store_block(blk, true)?;
+
+                // Accept block transactions into the VM
+                if blk.header.iteration == 1 {
+                    return vm.finalize(blk);
+                }
+
+                //TODO: Retrieve eligible_provisioners
+                vm.accept(blk)
+            })
+        }?;
```

Comment:
Persist block should be done only after ensure that state_root returned by `accept`/`finalize` is the same of the block header

See also #924